src/extension.rs

```diff
@@ -175,10 +175,22 @@ impl ExtensionSequence {
             while offset < len {
                 let esize = source.read_i32()?;
                 let ecode = source.read_i32()?;
-                let data_size = esize as usize - 8;
-                let mut edata = vec![0u8; data_size];
-                source.read_exact(&mut edata)?;
-                extensions.push(Extension::new(esize, ecode, edata));
+
+                let data_size = (esize as usize).saturating_sub(8);
+                // rather than pre-allocating for the data size, this will
+                // pre-allocate up to a more reliable amount and feed the
+                // vector sequentially, to prevent some trivial OOM attacks
+                let mut edata = Vec::with_capacity(usize::min(data_size, 1 << 25 /* 32M */));
```

Comment:
I just wrote this unconsciously

Reply:
I just wrote like this unconsciously: `min` is a symmetrical operation, and I didn't feel like there had to be a _receiving_ member. But yes, I don't have a strong opinion on this, and `a.min(b)` is shorter.

Reply:
As for the constants, these are mostly "magic" numbers obtained by intuition. It just has to be something large enough to prevent too many unnecessary allocations, but small enough to avoid an immediate OOM. I suppose I can move them to constant bindings for a better code organization.